README.md
# raerdata

raerdata provides external  data sets for calling rna-editing sites with [raer](https://rnabioco.github.io/raer).

## Installation

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("raerdata")
```

## To use `raerdata`

``` r
library(ExperimentHub)
eh <- ExperimentHub()

## query
refs <- query(eh, "raerdata")
refs

```


NEWS
Changes in version 0.99.0 (2023-05-18)
+ init


R/raerdata.R
#' raerdata
#'
#' @description A collection of datasets and databases to demonstrate identification of RNA-editing sites using the raer package.
#'
#' @docType package
#' @keywords internal
#' @importFrom ExperimentHub ExperimentHub
#' @importFrom utils read.csv
#' @rdname raerdata
#' @name raerdata
NULL

#' Databases of known RNA editing sites
#'
#'
#' @details
#' `rediportal_full_hg38()` will download the human rediportal database for hg38 which has been
#'  coverted into a GRanges object. The Granges is supplemented with additional columns of information
#'  provided by the rediportal database, including gene location, repeat type, dbSNP annotation, and
#'  potential for amino-acid recoding.
#'
#' `rediportal_coords_hg38()` will download the human rediportal database for hg38 which has been
#'  coverted into a GRanges object, which only contains the coordinates of the editing site.
#'
#' `rediportal_full_mm10()` will download the mouse rediportal database for mm10 which has been
#'  coverted into a GRanges object. The Granges is supplemented with additional columns of information
#'  provided by the rediportal database, including gene location, repeat type, dbSNP annotation, and
#'  potential for amino-acid recoding.
#'
#' `rediportal_coords_mm10()`will download the mouse rediportal database for mm10 which has been
#'  coverted into a GRanges object, which only contains the coordinates of the editing site.
#'
#'  `gabay_sites_hg38()` will download high-confidence human CDS editing sites (hg38).
#'
#'  `gabay_sites_mm10()` will download high-confidence mouse CDS editing sites (lifted-over from hg38 to mm10).
#'
#' @rdname atlases
#' @examples
#' rediportal_full_mm10()
NULL

#' @rdname rediportal
"rediportal_full_mm10"
#' @rdname rediportal
"gabay_sites_hg38"
#' @rdname rediportal
"rediportal_coords_mm10"


#' Whole genome and RNA sequencing data from NA12877 cell line
#'
#'
#' @details
#' `NA12877_wgs()` will download a BAM and BAM index file from whole genome sequencing
#'  of the NA12877 cell line.
#'
#' `NA12877_rnaseq()` will download a BAM and BAM index file from RNA-seq
#'  of the NA12877 cell line.
#' @returns A data.frame containing:
#' \itemize{
#' \item \code{Name}, the name of the BAM file without the \code{*.bam} extension.
#' \item \code{Description}, a short string containing the experimental condition and replicate number.
#' \item \code{Path}, a \linkS4class{BamFile} containing the path to each BAM file.
#'
#' @rdname NA12877
#'
NULL

#' RNA sequencing data from study GSE99249
#'
#' @description
#' Study [GSE99249](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249) examined RNA editing in
#' WT, ADAR1KO, and ADAR1-p150 HEK293T cells treated with and without Interferon Beta.
#'
#' @details
#' `GSE99249()` will download a BAM and BAM index files from 6 RNA-seq libraries
#'
#' @returns A data.frame containing:
#' \itemize{
#' \item \code{Name}, the name of the BAM file without the \code{*.bam} extension.
#' \item \code{Description}, a short string containing the experimental condition and replicate number.
#' \item \code{Path}, a \linkS4class{BamFile} containing the path to each BAM file.
#'
#' @rdname GSE99249
#'
NULL

#' RNA sequencing data from study GSE99249
#'
#' @description
#' Study [GSE99249](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249) examined RNA editing in
#' WT, ADAR1KO, and ADAR1-p150 HEK293T cells treated with and without Interferon Beta.
#'
#' @details
#' `GSE99249()` will download a BAM and BAM index files from 6 RNA-seq libraries
#'
#' @returns A data.frame containing:
#' \itemize{
#' \item \code{Name}, the name of the BAM file without the \code{*.bam} extension.
#' \item \code{Description}, a short string containing the experimental condition and replicate number.
#' \item \code{Path}, a \linkS4class{BamFile} containing the path to each BAM file.
#'
#' @rdname GSE99249
#'
NULL

#' RNA sequencing data from study GSE99249
#'
#' @description
#' Study [GSE99249](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249) examined RNA editing in
#' WT, ADAR1KO, and ADAR1-p150 HEK293T cells treated with and without Interferon Beta.
#'
#' @details
#' `GSE99249()` will download a BAM and BAM index files from 6 RNA-seq libraries
#'
#' @returns A data.frame containing:
#' \itemize{
#' \item \code{Name}, the name of the BAM file without the \code{*.bam} extension.
#' \item \code{Description}, a short string containing the experimental condition and replicate number.
#' \item \code{Path}, a \linkS4class{BamFile} containing the path to each BAM file.
#'
#' @rdname GSE99249
#'
NULL


inst/scripts/make-metadata.R
library(here)
rediportal <- data.frame(
    Title = c(
        "Rediportal RNA editing sites, all data, (mm10)",
        "Rediportal RNA editing sites, coordinates only, (mm10)",
        "Rediportal RNA editing sites, all data, (hg38)",
        "Rediportal RNA editing sites, coordinates only, (hg38)",
        "Mouse adenosine-to-inosine RNA recoding sites",
        "Human adenosine-to-inosine RNA recoding sites"
    ),
    Description = c(
        "GRanges object containing 107,094 mouse RNA editing sites, mm10, additional columns describe editing site characteristics, as described by REDIPORTAL http://srv00.recas.ba.infn.it/atlas/help.html",
        "GRanges object containing 107,094 mouse RNA editing sites",
        "GRanges object containing 15,638,648 human RNA editing sites, hg38, additional columns describe editing site characteristics, as described by REDIPORTAL http://srv00.recas.ba.infn.it/atlas/help.html",
        "GRanges object containing 15,638,648 human RNA editing sites, hg38",
        "GRanges object containing 1,160 mouse CDS recoding RNA editing sites, mm10",
        "GRanges object containing 1,517 human CDS recoding RNA editing sites, hg38"
    ),
    RDataPath = file.path("raerdata",
                          "1.0.0",
                          c(
                              "rediportal_full_mm10.rda",
                              "rediportal_coords_mm10.rda",
                              "rediportal_full_hg38.rda",
                              "rediportal_coords_hg38.rda",
                              "gabay_sites_mm10.rds",
                              "gabay_sites_hg38.rds"
                           )),
    BiocVersion="3.18",
    Genome=c(
        "mm10",
        "mm10",
        "hg38",
        "hg38",
        "mm10",
        "hg38"
    ),
    SourceType="RDA",
    SourceUrl=c(
        "http://srv00.recas.ba.infn.it/atlas/",
        "http://srv00.recas.ba.infn.it/atlas/",
        "http://srv00.recas.ba.infn.it/atlas/",
        "http://srv00.recas.ba.infn.it/atlas/",
        "https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-022-28841-4/MediaObjects/41467_2022_28841_MOESM3_ESM.xlsx",
        "https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-022-28841-4/MediaObjects/41467_2022_28841_MOESM3_ESM.xlsx"
  ),
    SourceVersion="1",
    Species=c(
        "Mus musculus",
        "Mus musculus",
        "Homo sapiens",
        "Homo sapiens",
        "Mus musculus",
        "Homo sapiens"
    ),
    TaxonomyId=c(
        "10090",
        "10090",
        "9606",
        "9606",
        "10090",
        "9606"
    ),
    Coordinate_1_based = TRUE,
    DataProvider=c(
        "REDIportal",
        "REDIportal",
        "REDIportal",
        "REDIportal",
        "https://doi.org/10.1038/s41467-022-28841-4",
        "https://doi.org/10.1038/s41467-022-28841-4"
    ),
    Maintainer="Kent Riemondy <kent.riemondy@cuanschutz.com>",
    RDataClass="GRanges",
    DispatchClass="Rda",
    stringsAsFactors = FALSE
)

adar_ko <- data.frame(

)

sc_data <- data.frame(

)

rna_dna <- data.frame(

)

misaligned_sites <- data.frame(

)

main.data <- rbind(rediportal, adar_ko, pbmc_sc, rna_dna, misaligned_sites)
write.csv(file=here("inst/extdata/metadata.csv"), main.data, row.names=FALSE)


inst/scripts/dbases/get_databases.sh
#! /usr/bin/env bash

# download genome sequences and transcript annotations
#wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/GRCh38.primary_assembly.genome.fa.gz
#gunzip GRCh38.primary_assembly.genome.fa.gz
#samtools faidx GRCh38.primary_assembly.genome.fa
#
#wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/GRCm38.primary_assembly.genome.fa.gz
#gunzip GRCm38.primary_assembly.genome.fa.gz
#samtools faidx GRCm38.primary_assembly.genome.fa
#
#wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/gencode.v37.annotation.gtf.gz
#gunzip gencode.v37.annotation.gtf.gz
#
#wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz
#gunzip gencode.vM25.annotation.gtf.gz

# make STAR index
STAR \
  --runMode genomeGenerate  \
  --runThreadN 12 \
  --genomeDir star/GRCm38 \
  --genomeFastaFiles GRCm38.primary_assembly.genome.fa


inst/scripts/rediportal/get_rediportal.R
library(data.table)
library(here)
library(tools)
library(Rsamtools)
library(GenomicRanges)
library(readxl)
library(AnnotationHub)
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Mmusculus.UCSC.mm10)
library(rtracklayer)

# Known editing sites from the Rediportal database:
# Mansi L, Tangaro MA, Lo Giudice C, et al. REDIportal: millions of novel A-to-I RNA editing events from thousands of RNAseq experiments. Nucleic Acids Res [Internet]. 2021;49:D1012–D1019. Available from: http://dx.doi.org/10.1093/nar/gkaa916.

data_dir <- here("inst/scripts/rediportal")

# hg38 database of known RNA editing sites from Rediportal (v2)
url <- "http://srv00.recas.ba.infn.it/webshare/ATLAS/donwload/TABLE1_hg38.txt.gz"

rediportal_flat_file <- file.path(data_dir, "TABLE1_hg38.txt.gz")
if(!file.exists(rediportal_flat_file)){
    download.file(url, rediportal_flat_file)
    md5 <- md5sum(rediportal_flat_file)
    stopifnot(unname(md5) == "cd809b839ccb38d1b6b60184a2d190e2")
}

dat <- fread(rediportal_flat_file, data.table = FALSE)
dat <- dat[order(dat$Region, dat$Position), ]

gr <- makeGRangesFromDataFrame(dat,
                               seqnames.field = "Region",
                               start.field = "Position",
                               end.field = "Position",
                               strand.field = "Strand",
                               keep.extra.columns = TRUE)
genome(gr) <- "hg38"

# double check coordinates are correct (should be all A bases w.r.t strand)
seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, gr)

# 43,223 sites are not A in reference, not sure why.
gr <- gr[seqs == "A"]
gr_simple <- gr
mcols(gr_simple) <- NULL

save(gr, file = file.path(data_dir, "rediportal_full_hg38.rda"))
save(gr_simple, file = file.path(data_dir, "rediportal_coords_hg38.rda"))


#mm10 database of known RNA editing sites from Rediportal (v2)
url <-  "http://srv00.recas.ba.infn.it/webshare/ATLAS/donwload//TABLE1_mm10.txt.gz"

rediportal_flat_file <- file.path(data_dir, "TABLE1_mm10.txt.gz")
if(!file.exists(rediportal_flat_file)){
    download.file(url, rediportal_flat_file)
    md5 <- md5sum(rediportal_flat_file)
    stopifnot(unname(md5) == "4f0e8a3b47cee70f7bab855236e4c4e4")
}

dat <- fread(rediportal_flat_file, data.table = FALSE)
dat <- dat[order(dat$Region, dat$Position), ]
gr <- makeGRangesFromDataFrame(dat,
                               seqnames.field = "Region",
                               start.field = "Position",
                               end.field = "Position",
                               strand.field = "Strand",
                               keep.extra.columns = TRUE)
genome(gr) <- "mm10"

# double check coordinates are correct (should be all A bases w.r.t strand)
seqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, gr)

# all sites are A in reference, unnecessary, but may be needed in future versions
gr <- gr[seqs == "A"]
gr_simple <- gr
mcols(gr_simple) <- NULL

save(gr, file = file.path(data_dir, "rediportal_full_mm10.rda"))
save(gr_simple, file = file.path(data_dir, "rediportal_coords_mm10.rda"))

# Sites from:
# Gabay O, Shoshan Y, Kopel E, et al. Landscape of adenosine-to-inosine RNA recoding across human tissues. Nat Commun [Internet]. 2022;13:1184. Available from: http://dx.doi.org/10.1038/s41467-022-28841-4.

if(!file.exists(file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"))){
    download.file("https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-022-28841-4/MediaObjects/41467_2022_28841_MOESM3_ESM.xlsx",
                  file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"))
}

gabay_sites <- read_excel(file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"),
                          sheet = 3)
gabay_sites <- makeGRangesFromDataFrame(gabay_sites,
                             seqnames = "Chromosome",
                             start.field = "Postion",
                             end.field = "Postion",
                             strand.field = "Strand",
                             keep.extra.columns = TRUE)
genome(gabay_sites) <- "hg38"
seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, gabay_sites)
# all sites are A in hg38
stopifnot(all(seqs == "A"))
save(gabay_sites, file = file.path(data_dir, "gabay_sites_hg38.rda"))

# liftover to mm10
ah <- AnnotationHub()
chainfile <- ah[["AH14109"]]
mm10_gabay_sites <- liftOver(gabay_sites, chainfile)
# 1448 of 1517 hg38 sites have 1-to-1 mapping,
mm10_gabay_sites <- unlist(mm10_gabay_sites[elementNROWS(mm10_gabay_sites) == 1])

# ensure that site is an A in mm10
seqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, mm10_gabay_sites)
# 1160 of 1148 are A
mm10_gabay_sites <- mm10_gabay_sites[seqs == "A"]
genome(gabay_sites) <- "mm10"

save(mm10_gabay_sites, file = file.path(data_dir, "gabay_sites_mm10.rda"))


inst/scripts/rediportal-data-processing.R
library(data.table)
library(here)
library(tools)
library(Rsamtools)
library(GenomicRanges)
library(readxl)
library(AnnotationHub)
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Mmusculus.UCSC.mm10)
library(rtracklayer)

# Known editing sites from the Rediportal database:
# Mansi L, Tangaro MA, Lo Giudice C, et al. REDIportal: millions of novel A-to-I RNA editing events from thousands of RNAseq experiments. Nucleic Acids Res [Internet]. 2021;49:D1012–D1019. Available from: http://dx.doi.org/10.1093/nar/gkaa916.

data_dir <- here("inst/scripts/rediportal")
dir.create(data_dir, showWarnings = FALSE)

# hg38 database of known RNA editing sites from Rediportal (v2)
url <- "http://srv00.recas.ba.infn.it/webshare/ATLAS/donwload/TABLE1_hg38.txt.gz"

rediportal_flat_file <- file.path(data_dir, "TABLE1_hg38.txt.gz")
if(!file.exists(rediportal_flat_file)){
    download.file(url, rediportal_flat_file)
    md5 <- md5sum(rediportal_flat_file)
    stopifnot(unname(md5) == "cd809b839ccb38d1b6b60184a2d190e2")
}

dat <- fread(rediportal_flat_file, data.table = FALSE)
dat <- dat[order(dat$Region, dat$Position), ]

gr <- makeGRangesFromDataFrame(dat,
                               seqnames.field = "Region",
                               start.field = "Position",
                               end.field = "Position",
                               strand.field = "Strand",
                               keep.extra.columns = TRUE)
genome(gr) <- "hg38"

# double check coordinates are correct (should be all A bases w.r.t strand)
seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, gr)

# 43,223 sites are not A in reference, not sure why.
gr <- gr[seqs == "A"]
gr_simple <- gr
mcols(gr_simple) <- NULL

save(gr, file = file.path(data_dir, "rediportal_full_hg38.rda"))
save(gr_simple, file = file.path(data_dir, "rediportal_coords_hg38.rda"))


#mm10 database of known RNA editing sites from Rediportal (v2)
url <-  "http://srv00.recas.ba.infn.it/webshare/ATLAS/donwload//TABLE1_mm10.txt.gz"

rediportal_flat_file <- file.path(data_dir, "TABLE1_mm10.txt.gz")
if(!file.exists(rediportal_flat_file)){
    download.file(url, rediportal_flat_file)
    md5 <- md5sum(rediportal_flat_file)
    stopifnot(unname(md5) == "4f0e8a3b47cee70f7bab855236e4c4e4")
}

dat <- fread(rediportal_flat_file, data.table = FALSE)
dat <- dat[order(dat$Region, dat$Position), ]
gr <- makeGRangesFromDataFrame(dat,
                               seqnames.field = "Region",
                               start.field = "Position",
                               end.field = "Position",
                               strand.field = "Strand",
                               keep.extra.columns = TRUE)
genome(gr) <- "mm10"

# double check coordinates are correct (should be all A bases w.r.t strand)
seqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, gr)

# all sites are A in reference, unnecessary, but may be needed in future versions
gr <- gr[seqs == "A"]
gr_simple <- gr
mcols(gr_simple) <- NULL

save(gr, file = file.path(data_dir, "rediportal_full_mm10.rda"))
save(gr_simple, file = file.path(data_dir, "rediportal_coords_mm10.rda"))

# Sites from:
# Gabay O, Shoshan Y, Kopel E, et al. Landscape of adenosine-to-inosine RNA recoding across human tissues. Nat Commun [Internet]. 2022;13:1184. Available from: http://dx.doi.org/10.1038/s41467-022-28841-4.

if(!file.exists(file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"))){
    download.file("https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-022-28841-4/MediaObjects/41467_2022_28841_MOESM3_ESM.xlsx",
                  file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"))
}

gabay_sites <- read_excel(file.path(data_dir, "gabay_et_al_2022_sup_files.xlsx"),
                          sheet = 3)
gabay_sites <- makeGRangesFromDataFrame(gabay_sites,
                             seqnames = "Chromosome",
                             start.field = "Postion",
                             end.field = "Postion",
                             strand.field = "Strand",
                             keep.extra.columns = TRUE)
genome(gabay_sites) <- "hg38"
seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, gabay_sites)
# all sites are A in hg38
stopifnot(all(seqs == "A"))
save(gabay_sites, file = file.path(data_dir, "gabay_sites_hg38.rda"))

# liftover to mm10
ah <- AnnotationHub()
chainfile <- ah[["AH14109"]]
mm10_gabay_sites <- liftOver(gabay_sites, chainfile)
# 1448 of 1517 hg38 sites have 1-to-1 mapping,
mm10_gabay_sites <- unlist(mm10_gabay_sites[elementNROWS(mm10_gabay_sites) == 1])

# ensure that site is an A in mm10
seqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, mm10_gabay_sites)
# 1160 of 1148 are A
mm10_gabay_sites <- mm10_gabay_sites[seqs == "A"]
genome(gabay_sites) <- "mm10"

save(mm10_gabay_sites, file = file.path(data_dir, "gabay_sites_mm10.rda"))


inst/scripts/mismapped/blat/blat_mismatches.sh

#BSUB -n 44
#BSUB -J align 
#BSUB -o log_%J.out
#BSUB -e log_%J.err
#BSUB -R "select[mem>50] rusage[mem=50] span[hosts=1]"
#BSUB -q rna

pblat \
    ../../dbases/GRCh38.primary_assembly.genome.fa \
    GRCm38.75.fasta \
    -threads=44 \
    -stepSize=5 \
    -repMatch=2253 \
    -minScore=0 \
    -minIdentity=0 \
    -out=pslx \
    -noHead \
    GRCm38.75.pslx

# code from gabay et al to parse blat output...
in_pslx=GRCm38.75.pslx
awk '$1>=68 && $2>0 && $5==0 {aln_blocks_sum=0; split($19,aln_blocks,","); for (i=1;i<=$18;i++) aln_blocks_sum+=aln_blocks[i]; if ($1/aln_blocks_sum >= 0.96) print $0}' $in_pslx \
    | awk 'BEGIN{FS=OFS="\t"; rev["A"]="T";rev["C"]="G";rev["G"]="C";rev["T"]="A"}{aln_length=$1; aln_strand=$9; n=split($10,query_name,"_"); query_chr=query_name[4]; query_strand=query_name[3]; if (query_strand=="-") query_strand_flag=1; else query_strand_flag=0; j=1; for (i=5; i<=n; i++) {split(query_name[i],query_intervals,"-"); query_intervals[1]+=1; 
        for (w=query_intervals[1+(1*query_strand_flag)];
        w!=query_intervals[2-(1*query_strand_flag)]+(1-(2*query_strand_flag));w=w+(1-(2*query_strand_flag))){query_coords[j]=w;j++}}
        n_aln=$18; split($19,aln_blocks,","); split($20,query_blocks,",");
        split($21,target_blocks,","); split($22,query_seqs,",");
        split($23,target_seqs,","); if (aln_strand=="-")
        aln_strand_flag=1; else aln_strand_flag=0; 
        for (cur_block=n_aln-(n_aln+1)*aln_strand_flag;cur_block!=0+(n_aln+1)*aln_strand_flag;
        cur_block=cur_block-1+2*aln_strand_flag){cur_aln_length=aln_blocks[cur_block];
        for (pos=1; pos<=cur_aln_length;pos++)
            {query_nuc=toupper(substr(query_seqs[cur_block],pos,1));
                target_nuc=toupper(substr(target_seqs[cur_block],pos,1));
                if (query_nuc!=target_nuc) {
                    if (aln_strand=="-"){query_nuc=rev[query_nuc];target_nuc=rev[target_nuc]} query_pos=pos+query_blocks[cur_block];
                    query_coord=query_coords[query_pos+(j-1-2*query_pos+1)*aln_strand_flag];
                    target_pos=pos+target_blocks[cur_block]-1; 
                    print query_chr,query_coord-1,query_coord, query_nuc target_nuc,"0", query_strand, $14 "_" target_pos "_" target_pos+1 "_" aln_strand}}}}' \
                      


inst/scripts/misalignment-db-data-processing.sh
#BSUB -n 44
#BSUB -J align
#BSUB -o log_%J.out
#BSUB -e log_%J.err
#BSUB -R "select[mem>50] rusage[mem=50] span[hosts=1]"
#BSUB -q rna

# note requires bedtools, perl, and pblat in PATH
# uses ~44 CPU cores for processing

# Approach, awk code, and perl scripts are from https://github.com/a2iEditing/deNovo-Detect
# related to citation
# Gabay, O., Shoshan, Y., Kopel, E. et al. Landscape of adenosine-to-inosine RNA recoding across human tissues. Nat Commun 13, 1184 (2022). https://doi.org/10.1038/s41467-022-28841-4

# RefSeq_Curated_NM_CDS_010817.bed RefSeq_Curated_NM_Exons_010817.bed  RefSeq_Curated_NM_Gene_010817.bed
# were all downloaded from the table browser.
# RefSeq_Curated_NM_Gene_010817.bed was modified to match formatting for CDS and Exons using:
# awk 'BEGIN {OFS=FS="\t"} strand=($6 == "+")?"f":"r" {print $1,$2,$3,$4"_gene_0_0_"$1"_"$2 + 1"_"strand, $5, $6}' > RefSeq_Curated_NM_Gene_010817_mod.bed

mkdir -p misalignment_db
cd misalignment_db

if [[ ! -f "hg38_all.fa" ]]; then
  wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.analysisSet.chroms.tar.gz
  tar -xzvf hg38.analysisSet.chroms.tar.gz
  cat ./hg38.analysisSet.chroms/*.fa > hg38_all.fa
  rm hg38.analysisSet.chroms.tar.gz
  rm -rf hg38.analysisSet.chroms/
fi

get_misalignment_db() {
    in_bed=$1
    fa=$2
    pslx=$3
    out_bed=$4
    threads=$5
    bedtools getfasta -bed $in_bed -fi $fa -s -bedOut 2> /dev/null \
      | awk 'BEGIN{FS=OFS="\t"}{split($4,a,"_"); print $1,$2,$3,a[1]"_"a[2],$5,$6,$7}' \
      | grep -i -v "NN" \
      | perl sequence_splicer_gene_coords.pl /dev/stdin tmp.fa


    perl sequence_chopper_with_coords.pl \
      tmp.fa \
      76 \
      19 \
      tmp_chopped.fa

    # setting minScore and minIdentity to 0 seems unnecessary if we will post filter on match length >= 68
    # changing to defaults of 30 and 90 respectfully
    pblat \
        $fa \
        tmp_chopped.fa \
        -threads=$threads \
        -stepSize=5 \
        -repMatch=2253 \
        -minScore=30 \
        -minIdentity=90 \
        -out=pslx \
        -noHead \
        $pslx


    awk '$1>=68 && $2>0 && $5==0 {aln_blocks_sum=0; split($19,aln_blocks,",");
        for (i=1;i<=$18;i++) aln_blocks_sum+=aln_blocks[i];
          if ($1/aln_blocks_sum >= 0.96) print $0}' $pslx \
        | awk 'BEGIN{FS=OFS="\t";
                     rev["A"]="T";
                     rev["C"]="G";
                     rev["G"]="C";
                     rev["T"]="A"}
               {aln_length=$1;
                aln_strand=$9;
                n=split($10,query_name,"_");
                query_chr=query_name[4];
                query_strand=query_name[3];

                if (query_strand=="-") query_strand_flag=1;
                else query_strand_flag=0;
                j=1;

                for (i=5; i<=n; i++) {split(query_name[i], query_intervals, "-");
                  query_intervals[1]+=1;
                  for (w=query_intervals[1+(1*query_strand_flag)];
                    w!=query_intervals[2-(1*query_strand_flag)]+(1-(2*query_strand_flag));
                    w=w+(1-(2*query_strand_flag))) {query_coords[j]=w;j++}
                }
                n_aln=$18;
                split($19,aln_blocks,",");
                split($20,query_blocks,",");
                split($21,target_blocks,",");
                split($22,query_seqs,",");
                split($23,target_seqs,",");
                if (aln_strand=="-")
                  aln_strand_flag=1;
                else
                  aln_strand_flag=0;

                for (cur_block=n_aln-(n_aln+1)*aln_strand_flag;cur_block!=0+(n_aln+1)*aln_strand_flag;
                     cur_block=cur_block-1+2*aln_strand_flag){cur_aln_length=aln_blocks[cur_block];
                for (pos=1; pos<=cur_aln_length;pos++)
                    {query_nuc=toupper(substr(query_seqs[cur_block],pos,1));
                    target_nuc=toupper(substr(target_seqs[cur_block],pos,1));
                    if (query_nuc!=target_nuc) {
                        if (aln_strand=="-"){query_nuc=rev[query_nuc];target_nuc=rev[target_nuc]} query_pos=pos+query_blocks[cur_block];
                        query_coord=query_coords[query_pos+(j-1-2*query_pos+1)*aln_strand_flag];
                        target_pos=pos+target_blocks[cur_block]-1;
                        print query_chr, query_coord-1, query_coord, query_nuc target_nuc,"0", query_strand, $14 "_" target_pos "_" target_pos+1 "_" aln_strand}}}}' \
        > $out_bed

        rm -f tmp.fa tmp_chopped.fa $pslx
}

get_misalignment_db \
  RefSeq_Curated_NM_CDS_010817.bed \
  hg38_all.fa \
  RefSeq_Curated_NM_CDS_010817_spliced_noNs_chopped_76_19.pslx \
  RefSeq_Curated_NM_CDS_010817_spliced_noNs_chopped_76_19_mms.bed \
  44

get_misalignment_db \
  RefSeq_Curated_NM_Exons_010817.bed \
  hg38_all.fa \
  RefSeq_Curated_NM_Exons_010817_spliced_noNs_chopped_76_19.pslx \
  RefSeq_Curated_NM_Exons_010817_spliced_noNs_chopped_76_19_mms.bed \
  44

get_misalignment_db \
  RefSeq_Curated_NM_Gene_010817_mod.bed \
  hg38_all.fa \
  RefSeq_Curated_NM_Gene_010817_mod_spliced_noNs_chopped_76_19.pslx \
  RefSeq_Curated_NM_Gene_010817_mod_spliced_noNs_chopped_76_19_mms.bed \
  44
